service/qos/service_level_controller.cc

```diff
@@ -231,15 +231,30 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
 }
 
 future<>  service_level_controller::notify_service_level_added(sstring name, service_level sl_data) {
-    _service_levels_db.emplace(name, sl_data);
-    return make_ready_future();
+    return seastar::async( [this, name, sl_data] {
+        _subscribers.for_each([name, sl_data] (qos_configuration_change_subscriber* subscriber) {
+            subscriber->on_before_service_level_add(name, sl_data.slo).get();
+        });
+    }).then([this, name, sl_data] () {
```

Comment:
nitpick: you already created a thread, so you can just as well call `_service_levels_db.emplace(name, sl_data);` inside the thread as well, which saves us some lines and an allocation

Reply:
Also - perhaps we should consider what happens if one of the `on_before_service_level_add` callbacks throws - currently, we won't emplace a new service level to `_service_levels_db` - is that expected? Or should we just log the error and still try to emplace the service level? Same question applies to `on_before_service_level_change` below

Reply:
You are right, we should at least try to continue.